### PR TITLE
attempt to fix-2906

### PR DESCRIPTION
**Pull request by nishantmonu51**
Another way of fixing #2906. 
Refer discussion on  #2906 for more details

### CODE REVIEW

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java [hunk 1]
```diff
@@ -213,7 +213,7 @@ public static HadoopDruidIndexerConfig fromConfiguration(Configuration conf)
   private volatile HadoopIngestionSpec schema;
   private volatile PathSpec pathSpec;
   private volatile Map<DateTime, ShardSpecLookup> shardSpecLookups = Maps.newHashMap();
-  private volatile Map<ShardSpec, HadoopyShardSpec> hadoopShardSpecLookup = Maps.newHashMap();
+  private volatile Map<DateTime, Map<ShardSpec, HadoopyShardSpec>> hadoopShardSpecLookup = Maps.newHashMap();
```
Comment (gianm) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: this and `shardSpecLookups` could be final

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java [hunk 2]
```diff
@@ -242,9 +242,13 @@ public ShardSpec apply(HadoopyShardSpec input)
               )
           )
       );
+
+      Map<ShardSpec, HadoopyShardSpec> hadoopyShardSpecLookup = Maps.newHashMap();
```
Comment (gianm) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: suggest rename to `innerHadoopShardSpecLookup` – "hadoopy" is easy to mistake for "hadoop" when someone is reading this code

**Comment by gianm** [comment]
👍 other than minor comments

**Comment by gianm** [comment]
Looks like this bug has been around for a long time, but the behavior from #2906 was new in 0.9.0 due to https://github.com/druid-io/druid-api/commit/95b5557583f38423a3c0a3f1504795a3bdeed917. Beforehand this bug was often masked because different NoneShardSpec instances, even though they were all equal, would often have different hashCodes and so the map could store a few of them without them colliding.

**Comment by gianm** [comment]
@nishantmonu51 btw, is it possible to have a test for this?

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java (hunk 2 above)
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: also suggest `final`

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java (hunk 1 above)
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: Agreed, and probably remove `volatile`

File: indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java [hunk 3]
```diff
@@ -393,18 +397,18 @@ public Job addInputPaths(Job job) throws IOException
     if (!timeBucket.isPresent()) {
       return Optional.absent();
     }
-
-    final ShardSpec actualSpec = shardSpecLookups.get(timeBucket.get().getStart())
+    DateTime bucketStart = timeBucket.get().getStart();
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java]: `final`

**Comment by nishantmonu51** [comment]
@gianm @drcrallen: Added test and handled review comments

**Comment by himanshug** [comment]
:+1:

**Comment by gianm** [comment]
👍

**Pull request merged**